components/collections/codepointtrie_builder/src/native.rs

```diff
@@ -0,0 +1,190 @@
+// don't want it to bother the linker. In that case, we leave this unused, and silence the warning
+#[allow(unused)]
+pub(crate) fn run_native<T>(cpt_builder: &CodePointTrieBuilder<T>) -> CodePointTrie<'static, T>
+where
+    T: TrieValue + Into<u32>,
+{
+    let mut error = 0;
+    let builder = unsafe {
+        // safety: we're passing a valid error pointer
+        // leak-safety: we clean up `builder` except in panicky codepaths
+        umutablecptrie_open(
+            cpt_builder.default_value.into(),
+            cpt_builder.error_value.into(),
+            &mut error,
+        )
+    };
+
+    if error != 0 {
+        panic!("cpt builder returned error code {}", error);
+    }
+
+    let CodePointTrieBuilderData::ValuesByCodePoint(values) = cpt_builder.data;
+
+    for (cp, value) in values.iter().enumerate() {
+        unsafe {
+            // safety: builder is a valid UMutableCPTrie
+            // safety: we're passing a valid error pointer
+            umutablecptrie_set(builder, cp as u32, (*value).into(), &mut error);
+        }
+        if error != 0 {
+            panic!("cpt builder returned error code {}", error);
+        }
+    }
+
+    let trie_type = match cpt_builder.trie_type {
+        TrieType::Fast => 0,
+        TrieType::Small => 1,
+    };
+    let width = match mem::size_of::<T::ULE>() {
+        2 => 0, // UCPTRIE_VALUE_BITS_16
+        4 => 1, // UCPTRIE_VALUE_BITS_32
+        1 => 2, // UCPTRIE_VALUE_BITS_8
+        other => panic!("Don't know how to make trie with width {other}"),
+    };
+    // safety: `builder` is a valid UMutableCPTrie
+    // safety: we're passing a valid error pointer
+    // leak-safety: we clean up `built` except in panicky codepaths
+    let built = unsafe { umutablecptrie_buildImmutable(builder, trie_type, width, &mut error) };
+    if error != 0 {
+        panic!("cpt builder returned error code {}", error);
+    }
+    unsafe {
+        // safety: builder is a valid UMutableCPTrie
+        // safety: we don't use builder after this
+        umutablecptrie_close(builder);
+    }
+
+    // safety: this is is a valid trie returned by umutablecptrie_buildImmutable (which did not evaluate to an error)
+    let trie = unsafe { &*built };
+
+    let header = CodePointTrieHeader {
+        high_start: trie.highStart,
+        shifted12_high_start: trie.shifted12HighStart,
+        index3_null_offset: trie.index3NullOffset,
+        data_null_offset: trie
+            .dataNullOffset
+            .try_into()
+            .expect("Found negative or out of range dataNullOffset"),
+        null_value: trie.nullValue,
+        trie_type: TrieType::try_from(trie.type_ as u8).expect("Found out of range TrieType"),
+    };
+
+    // safety: we expect ICU4C to give us a valid slice (index, indexLength). The pointer types
+    // are already strongly typed, giving the right slice type.
+    let index_slice = unsafe { slice::from_raw_parts(trie.index, trie.indexLength as usize) };
```

Comment:
Annoying safety issue, here and elsewhere: We should use `usize::try_from(x)` instead of `x as usize` and panic if it fails